File: src/util/hash.cpp
#include "flux/util/hash.hpp"
#include <array>
#include <fmt/format.h>
#include <openssl/evp.h>
#include <stdexcept>

namespace flux {

namespace {
Bytes compute_evp(const EVP_MD *type, std::span<const uint8_t> data) {
  if (!type) {
    throw std::runtime_error("Unknown hash algorithm");
  }

  unsigned int md_len = EVP_MD_size(type);
  Bytes hash(md_len);

  EVP_MD_CTX *ctx = EVP_MD_CTX_new();
  if (!ctx) {
    throw std::runtime_error("Failed to create EVP_MD_CTX");
  }

  if (EVP_DigestInit_ex(ctx, type, nullptr) != 1 ||
      EVP_DigestUpdate(ctx, data.data(), data.size()) != 1 ||
      EVP_DigestFinal_ex(ctx, hash.data(), &md_len) != 1) {
    EVP_MD_CTX_free(ctx);
    throw std::runtime_error("OpenSSL hashing failed");
  }

  EVP_MD_CTX_free(ctx);
  return hash;
}
} // namespace

Bytes Hash::compute(HashAlgorithm algo, std::span<const uint8_t> data) {
  switch (algo) {
  case HashAlgorithm::SHA1:
    return compute_sha1(data);
  case HashAlgorithm::SHA256:
    return compute_sha256(data);
  case HashAlgorithm::SHA3_256:
    return compute_sha3_256(data);
  case HashAlgorithm::BLAKE3:
    // BLAKE3 not standard in OpenSSL yet, fallback to SHA256 or throw?
    // Existing code fell back to SHA256. Let's keep fallback but maybe warn?
    // For now, consistent with previous behavior:
    return compute_sha256(data);
  }
  return compute_sha256(data);
}

Bytes Hash::compute_sha1(std::span<const uint8_t> data) {
  return compute_evp(EVP_sha1(), data);
}

Bytes Hash::compute_sha256(std::span<const uint8_t> data) {
  return compute_evp(EVP_sha256(), data);
}

Bytes Hash::compute_sha3_256(std::span<const uint8_t> data) {
  return compute_evp(EVP_sha3_256(), data);
}

size_t Hash::hash_size(HashAlgorithm algo) {
  switch (algo) {
  case HashAlgorithm::SHA1:
    return 20;
  case HashAlgorithm::SHA256:
    return 32;
  case HashAlgorithm::SHA3_256:
    return 32;
  case HashAlgorithm::BLAKE3:
    return 32;
  }
  return 32; // Default
}

std::string Hash::to_hex(std::span<const uint8_t> data) {
  static constexpr char hex_digits[] = "0123456789abcdef";
  std::string result;
  result.reserve(data.size() * 2);

  for (uint8_t byte : data) {
    result.push_back(hex_digits[byte >> 4]);
    result.push_back(hex_digits[byte & 0x0F]);
  }

  return result;
}

Result<Bytes> Hash::from_hex(std::string_view hex) {
  if (hex.size() % 2 != 0) {
    return flux::unexpected("Hex string must have even length");
  }

  Bytes result;
  result.reserve(hex.size() / 2);

  static constexpr int8_t lookup[] = {
      -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
      -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
      -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
      0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  -1, -1, -1, -1, -1, -1,
      -1, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1, -1, -1, -1,
      -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
      -1, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1, -1, -1, -1,
      -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1};

  for (size_t i = 0; i < hex.size(); i += 2) {
    char high = hex[i];
    char low = hex[i + 1];

    if (static_cast<unsigned char>(high) > 127 ||
        static_cast<unsigned char>(low) > 127 ||
        lookup[static_cast<unsigned char>(high)] == -1 ||
        lookup[static_cast<unsigned char>(low)] == -1) {
      return flux::unexpected(
          fmt::format("Invalid hex character at position {}", i));
    }

    result.push_back((lookup[static_cast<unsigned char>(high)] << 4) |
                     lookup[static_cast<unsigned char>(low)]);
  }

  return result;
}

} // namespace flux

File: src/util/hash.zig
const std = @import("std");
const types = @import("../core/types.zig");
const HashAlgorithm = types.HashAlgorithm;
const Bytes = types.Bytes;

// C bindings for OpenSSL (for SHA1 Git compatibility)
const c = @cImport({
    @cInclude("openssl/evp.h");
    @cInclude("openssl/sha.h");
});

/// Hash utilities for computing object IDs
pub const Hash = struct {
    /// Compute hash using specified algorithm
    pub fn compute(allocator: std.mem.Allocator, algo: HashAlgorithm, data: []const u8) ![]u8 {
        return switch (algo) {
            .sha1 => computeSha1(allocator, data),
            .sha256 => computeSha256(allocator, data),
            .sha3_256 => computeSha3_256(allocator, data),
            .blake3 => computeSha256(allocator, data), // Fallback to SHA256 for now
        };
    }

    /// Get hash size for algorithm
    pub fn hashSize(algo: HashAlgorithm) usize {
        return switch (algo) {
            .sha1 => 20,
            .sha256 => 32,
            .sha3_256 => 32,
            .blake3 => 32,
        };
    }

    /// Convert bytes to hex string
    pub fn toHex(allocator: std.mem.Allocator, data: []const u8) ![]u8 {
        const hex_chars = "0123456789abcdef";
        var result = try allocator.alloc(u8, data.len * 2);

        for (data, 0..) |byte, i| {
            result[i * 2] = hex_chars[byte >> 4];
            result[i * 2 + 1] = hex_chars[byte & 0x0F];
        }

        return result;
    }

    /// Convert hex string to bytes
    pub fn fromHex(allocator: std.mem.Allocator, hex: []const u8) ![]u8 {
        if (hex.len % 2 != 0) {
            return error.InvalidHash;
        }

        var result = try allocator.alloc(u8, hex.len / 2);
        errdefer allocator.free(result);

        for (0..result.len) |i| {
            const high = try hexCharToValue(hex[i * 2]);
            const low = try hexCharToValue(hex[i * 2 + 1]);
            result[i] = (high << 4) | low;
        }

        return result;
    }

    /// Compute SHA-1 hash (using OpenSSL for Git compatibility)
    pub fn computeSha1(allocator: std.mem.Allocator, data: []const u8) ![]u8 {
        const hash = try allocator.alloc(u8, 20);
        errdefer allocator.free(hash);

        const ctx = c.EVP_MD_CTX_new() orelse return error.OutOfMemory;
        defer c.EVP_MD_CTX_free(ctx);

        const sha1_type = c.EVP_sha1();
        if (c.EVP_DigestInit_ex(ctx, sha1_type, null) != 1) {
            return error.InvalidHash;
        }

        if (c.EVP_DigestUpdate(ctx, data.ptr, data.len) != 1) {
            return error.InvalidHash;
        }

        var md_len: c_uint = 20;
        if (c.EVP_DigestFinal_ex(ctx, hash.ptr, &md_len) != 1) {
            return error.InvalidHash;
        }

        return hash;
    }

    /// Compute SHA-256 hash (using Zig stdlib)
    fn computeSha256(allocator: std.mem.Allocator, data: []const u8) ![]u8 {
        var hash = try allocator.alloc(u8, 32);
        std.crypto.hash.sha2.Sha256.hash(data, hash[0..32], .{});
        return hash;
    }

    /// Compute SHA3-256 hash (using OpenSSL)
    fn computeSha3_256(allocator: std.mem.Allocator, data: []const u8) ![]u8 {
        const hash = try allocator.alloc(u8, 32);
        errdefer allocator.free(hash);

        const ctx = c.EVP_MD_CTX_new() orelse return error.OutOfMemory;
        defer c.EVP_MD_CTX_free(ctx);

        const sha3_type = c.EVP_sha3_256();
        if (c.EVP_DigestInit_ex(ctx, sha3_type, null) != 1) {
            return error.InvalidHash;
        }

        if (c.EVP_DigestUpdate(ctx, data.ptr, data.len) != 1) {
            return error.InvalidHash;
        }

        var md_len: c_uint = 32;
        if (c.EVP_DigestFinal_ex(ctx, hash.ptr, &md_len) != 1) {
            return error.InvalidHash;
        }

        return hash;
    }

    fn hexCharToValue(char: u8) !u8 {
        return switch (char) {
            '0'...'9' => char - '0',
            'a'...'f' => char - 'a' + 10,
            'A'...'F' => char - 'A' + 10,
            else => error.InvalidHash,
        };
    }
};

test "SHA-256 hash" {
    const allocator = std.testing.allocator;

    const data = "Hello, World!";
    const hash = try Hash.compute(allocator, .sha256, data);
    defer allocator.free(hash);

    try std.testing.expectEqual(@as(usize, 32), hash.len);
}

test "hex encoding/decoding" {
    const allocator = std.testing.allocator;

    const original = [_]u8{ 0xDE, 0xAD, 0xBE, 0xEF };
    const hex = try Hash.toHex(allocator, &original);
    defer allocator.free(hex);

    try std.testing.expectEqualStrings("deadbeef", hex);

    const decoded = try Hash.fromHex(allocator, hex);
    defer allocator.free(decoded);

    try std.testing.expectEqualSlices(u8, &original, decoded);
}

test "hash size" {
    try std.testing.expectEqual(@as(usize, 20), Hash.hashSize(.sha1));
    try std.testing.expectEqual(@as(usize, 32), Hash.hashSize(.sha256));
}
